### ISSUE TITLE
Sneak Attack doesn't give MDFC creatures haste

### ISSUE THREAD

**Issue by alexander-novo** [magefree/mage#8474]
[[Sneak Attack]] doesn't give MDFC creatures haste. Just observed with [[Tergrid, God of Fright]].

**Comment by weirddan455** [comment]
@jeffwadsworth I was also just taking a look at this.  It looks like MDFCs always enter the battlefield as their left half, which has a different UUID from the main card Id.  I have a fix but it's maybe not the best.  Check if it's an instance of ModalDoubleFacesCard.  If it is, get the permanent from the left side's Id.  There may be a cleaner way to fix this in engine code but I'm not sure.

Goryo's Vengeance and probably other cards are broken in the same way.

**Comment by JayDi85** [comment]
> enter the battlefield as their left half, which has a different UUID from the main card Id

That's correct. MDF cards and sides are fully independent cards with different ids (as example: transformable cards are same single cards with different characteristics).

There are dozen cards with same code style (move + gain haste). Search by:
* `new GainAbilityTargetEffect(HasteAbility.getInstance(), Duration.Custom);`
![shot_211123_002955](https://user-images.githubusercontent.com/8344157/142930870-d6444ef9-77f1-42d4-8725-ffbc554b3ec0.png)

So it must be improved for all cards.

Workaround 1 (different action order):
`GainAbilityTargetEffect` support targeting to cards and it used for same code style in other places (gives a haste before put card to battlefield). But it must add additional effect before move call.

Workaround 2 (usage of `CardUtil.getDefaultCardSideForBattlefield`):
Broken code:
```
controller.moveCards(card...
game.getPermanent(card...
```
Can be fixed by:
```
game.getPermanent(CardUtil.getDefaultCardSideForBattlefield(game, card).getId())
```

Maybe another fixes possible.

**Comment by JayDi85** [comment]
I added `ModalDoubleFacesCardsTest.test_FindMovedPermanentByCard` and fixed from workaround 2 in a5eb02515571727b53e95604f1db4d11090e4d52. So you can test, search other potential broken cards or find another workaround.

**Comment by JayDi85** [comment]
Too much cards with same code style (100+). Regexp searches:
* `moveCards.+\n.+game\.getPermanent`
* `moveCards.+\n.+\n.+game\.getPermanent`
* `moveCards.+\n.+\n.+\n.+game\.getPermanent`

There are comments in `Game->getPermanent` from commit 10cf9c4a4e03f709c9989d3299edb486ccb06ead:
> Find permanent on the battlefield by id. If you works with cards and want to check it on battlefield then use  getState().getZone() instead. Card's id and permanent's id can be different (example: mdf card puts half card to battlefield, not the main card).

There are other places with getPermanent methods:
`getBattlefield().getPermanent`
`getBattlefield().containsPermanent`

It can be modified to support mdfc search logic (e.g. find a half's permanent by main card id), but it's add a different logic to different getPermanent methods. So must be restricted.

**Conclusion**: moveCards + getPermanent combo must be improved to use `CardUtil.getDefaultCardSideForBattlefield(game, card)` instead.

**Comment by theelk801** [comment]
in general we want something better for tracking cards as they change zones, especially if we want to implement mutate properly
